src/rpc/connection.rs

```diff
@@ -693,3 +693,142 @@ impl CircularQueue {
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(_) => assert!(true),
+            Err(_) => assert!(false),
+        }
+    }
+
+    #[test]
+    fn test_get_metadata_ref() {
+        let mut queue = CircularQueue::new();
+        queue.init();
+        let mut recv_meta_data = vec![];
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(id) => match queue.get_meta_data_ref(id, recv_meta_data.len()) {
+                Ok(recv_meta_data_ref) => {
+                    assert_eq!(recv_meta_data_ref, &mut recv_meta_data);
+                }
+                Err(_) => assert!(false),
+            },
+            Err(_) => assert!(false),
+        }
+    }
+
+    #[test]
+    fn test_get_data_ref() {
+        let mut queue = CircularQueue::new();
+        queue.init();
+        let mut recv_meta_data = vec![];
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(id) => match queue.get_data_ref(id, recv_data.len()) {
+                Ok(recv_data_ref) => {
+                    assert_eq!(recv_data_ref, &mut recv_data);
+                }
+                Err(_) => assert!(false),
+            },
+            Err(_) => assert!(false),
+        }
+    }
+
+    #[test]
+    fn test_clean_up() {
+        let mut queue = CircularQueue::new();
+        queue.init();
+        let mut recv_meta_data = vec![];
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(id) => unsafe {
+                let callback = queue.callbacks[id as usize];
+                (*(callback as *mut OperationCallback)).state = CallbackState::Done;
+                match queue.clean_up() {
+                    Ok(_) => match (*(callback as *mut OperationCallback)).state {
+                        CallbackState::Empty => assert!(true),
+                        _ => assert!(false),
+                    },
+                    Err(_) => assert!(false),
+                }
+            },
+            Err(_) => assert!(false),
+        }
+    }
+
+    #[test]
+    fn test_wait_for_callback() {
+        let mut queue = CircularQueue::new();
+        queue.init();
+        let mut recv_meta_data = vec![];
+        let mut recv_data = vec![0u8; 1024];
+        let result = queue.register_callback(&mut recv_meta_data, &mut recv_data);
+        match result {
+            Ok(id) => {
+                let callback = queue.callbacks[id as usize];
+                unsafe {
+                    let oc = &*(callback as *mut OperationCallback);
+                    std::thread::spawn(|| match oc.channel.0.send(()) {
```

Comment:
If this new thread necessary?

Reply:
How about send signal in the same thread?

Reply:
I thought maps::channel might be synchronized......